api/end_device.proto

```diff
@@ -64,52 +64,52 @@ message MACParameters {
   // Downlink dwell time is set (400ms).
   bool downlink_dwell_time = 3;
   // ADR: data rate index to use.
-  DataRateIndex adr_data_rate_index = 4 [(gogoproto.customname) = "ADRDataRateIndex"];
+  DataRateIndex adr_data_rate_index = 4 [(gogoproto.customname) = "ADRDataRateIndex", (validate.rules).enum.defined_only = true];
   // ADR: transmission power index to use.
-  uint32 adr_tx_power_index = 5 [(gogoproto.customname) = "ADRTxPowerIndex"];
+  uint32 adr_tx_power_index = 5 [(gogoproto.customname) = "ADRTxPowerIndex", (validate.rules).uint32.lte = 15];
   // ADR: number of retransmissions.
-  uint32 adr_nb_trans = 6 [(gogoproto.customname) = "ADRNbTrans"];
+  uint32 adr_nb_trans = 6 [(gogoproto.customname) = "ADRNbTrans", (validate.rules).uint32.lte = 15];
   // ADR: number of messages to wait before setting ADRAckReq.
-  uint32 adr_ack_limit = 7 [(gogoproto.customname) = "ADRAckLimit"];
+  uint32 adr_ack_limit = 7 [(gogoproto.customname) = "ADRAckLimit", (validate.rules).uint32 = {gte: 1, lte: 32768}];
   // ADR: number of messages to wait after setting ADRAckReq and before changing TxPower or DataRate.
-  uint32 adr_ack_delay = 8 [(gogoproto.customname) = "ADRAckDelay"];
+  uint32 adr_ack_delay = 8 [(gogoproto.customname) = "ADRAckDelay", (validate.rules).uint32 = {gte: 1, lte: 32768}];
   // Rx1 delay (Rx2 delay is Rx1 delay + 1 second).
-  RxDelay rx1_delay = 9;
+  RxDelay rx1_delay = 9 [(validate.rules).enum.defined_only = true];
   // Data rate offset for Rx1.
-  uint32 rx1_data_rate_offset = 10;
+  uint32 rx1_data_rate_offset = 10 [(validate.rules).uint32.lte = 7];
   // Data rate index for Rx2.
-  DataRateIndex rx2_data_rate_index = 11;
+  DataRateIndex rx2_data_rate_index = 11 [(validate.rules).enum.defined_only = true];
   // Frequency for Rx2 (Hz).
-  uint64 rx2_frequency = 12;
+  uint64 rx2_frequency = 12 [(validate.rules).uint64 = {gte: 100000, lte: 1677721600}];
```

Comment:
Remove upper bound

Reply:
Why? This is the maximum value representable via MAC layer by LoRaWAN spec